dftimewolf/lib/enhancers/timesketch.py

```diff
@@ -0,0 +1,404 @@
+
+      # Move the datetime column to the first column displayed.
+      try:
+        index = columns.index('datetime')
+        datetime = columns.pop(index)
+        columns.insert(0, datetime)
+      except ValueError:
+        pass
+
+      self.state.StoreContainer(
+          containers.DataFrame(
+              data_frame=data_frame[columns],
+              name=self._REPORT_NAME,
+              description='Timesketch View: {0:s} - {1:s}'.format(
+                  view.name, view.description)))
+
+    return '\n'.join(view_strings)
+
+  def _WaitForAnalyzers(self, sketch):
+    """Wait for all analyzers to complete their run.
+
+    Args:
+      sketch (timesketch_api.sketch.Sketch): the sketch object.
+    """
+    check_number = 0
+    while True:
+      if check_number >= self._max_checks:
+        self.logger.warning(
+            'Exceeded maximum checks, not waiting any longer for analyzers '
+            'to complete.')
+        break
+
+      status_set = set()
+      # get_analyzer_status returns a dict with information about the run
+      # or all analyzers in a given sketch. One of the information is the
+      # current status of the analyzer run, ANALYZER_COMPLETE_SET contains
+      # the status values of analyzers that have completed their work.
+      for result in sketch.get_analyzer_status():
+        status_set.add(result.get('status', 'N/A'))
+
+      if status_set.issubset(self._ANALYZERS_COMPLETE_SET):
+        break
+
+      check_number += 1
+      time.sleep(self._ANALYZER_SECONDS_BETWEEN_CHECK)
+
+  def Process(self):
+    """Executes a Timesketch enhancer module."""
+    if self._timesketch_quick:
+      self.logger.info('Not waiting for analyzers to run.')
+      return
+
+    if not self.timesketch_api:
+      message = 'Could not connect to Timesketch server'
+      self.ModuleError(message, critical=True)
+
+    sketch = self.state.GetFromCache('timesketch_sketch')
+    if not sketch:
+      message = 'Missing the sketch, unable to connect to Timesketch.'
+      self.ModuleError(message, critical=True)
+
+    self.logger.info('Waiting for analyzers to complete their run.')
+
+    summary_lines = [self._formatter.Heading('Timesketch Run', level=1)]
+    summary_lines.append(self._formatter.Paragraph(
+        'This is a summary of actions taken by Timesketch '
+        'during its run.'))
+    summary_lines.append(self._formatter.Paragraph(
+        'To visit the sketch, click {0:s}'.format(self._formatter.Link(
+            url=self._GetSketchURL(sketch), text='here'))))
+    summary_lines.append(self._formatter.Paragraph(
+        'Here is an overview of actions taken:'))
+
+    self._WaitForAnalyzers(sketch)
+
+    # Force a refresh of sketch data.
+    _ = sketch.lazyload_data(refresh_cache=True)
+
+    summary_lines.append(self._formatter.IndentStart())
+
+    view_string = self._GenerateViewString(sketch)
+    if view_string:
+      summary_lines.append(self._formatter.IndentText(
```

Comment:
for easier debugging, please store the result of the `_formatter` into a variable and call `append` outside the `if` block. Same for other constructs like this below.

Reply:
ack